packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+            if (focusedIndexRef.current === -1) {
+              return dispatch([
+                'focus',
+                Number(optionsRef.current[length - 1].dataset.iuiIndex ?? -1),
+              ]);
             }
-            role='combobox'
-            aria-controls={isOpen ? `${id}-list` : undefined}
-            aria-autocomplete='list'
-            spellCheck={false}
-            autoCapitalize='none'
-            autoCorrect='off'
-            {...inputProps}
-            onChange={onInput}
-          />
-        </Popover>
-        <span
-          ref={toggleButtonRef}
-          className={cx('iui-end-icon', {
-            'iui-actionable': !inputProps?.disabled,
-            'iui-disabled': inputProps?.disabled,
-            'iui-open': isOpen,
-          })}
-          onClick={() => {
+
+            let prevIndex = focusedIndexRef.current;
+            do {
+              const currentElement = menuRef.current?.querySelector(
+                `[data-iui-index="${prevIndex}"]`,
+              );
+              const prevElement =
+                currentElement?.previousElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]:last-of-type');
+              prevIndex = Number(prevElement?.getAttribute('data-iui-index'));
+
+              if (prevElement?.ariaDisabled !== 'true') {
+                return dispatch(['focus', prevIndex]);
+              }
+            } while (prevIndex !== focusedIndexRef.current);
+            break;
+          }
+          case 'Enter': {
+            event.preventDefault();
             if (isOpen) {
-              setIsOpen(false);
+              dispatch(['select']);
             } else {
-              inputRef.current?.focus();
+              dispatch(['open']);
             }
-          }}
+            break;
+          }
+          case 'Escape': {
+            event.preventDefault();
+            dispatch(['close']);
+            break;
+          }
+          case 'Tab':
+            dispatch(['close']);
+            break;
+        }
+      },
+      [dispatch, isOpen, menuRef, onKeyDownProp, optionsRef],
+    );
+
+    const handleFocus = React.useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        dispatch(['open']);
+        onFocusProp?.(event);
+      },
+      [dispatch, onFocusProp],
+    );
+
+    return (
+      <Input
+        ref={refs}
+        onKeyDown={handleKeyDown}
+        onFocus={handleFocus}
+        aria-activedescendant={
+          isOpen && focusedIndex != undefined && focusedIndex > -1
+            ? getIdFromIndex(focusedIndex)
+            : undefined
+        }
+        role='combobox'
+        aria-controls={isOpen ? `${id}-list` : undefined}
+        aria-autocomplete='list'
+        spellCheck={false}
+        autoCapitalize='none'
+        autoCorrect='off'
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxInput.displayName = 'ComboBoxInput';
+
+const ComboBoxPopover = React.forwardRef(
```

Comment:
Maybe worth calling it `ComboBoxMenu` or `ComboBoxDropdown`

Reply:
Renamed to `ComboBoxDropdown`.

There is already a `ComboBoxMenu` expected as a child of this.